examples/aws-s3-folder/Main.scala

```diff
@@ -0,0 +1,88 @@
+import besom.*
+import besom.api.aws.*
+import besom.api.aws.s3.BucketObject
+import besom.api.aws.s3.inputs.BucketWebsiteArgs
+import besom.types.Asset.FileAsset
+import besom.types.Output
+import spray.json.*
+
+import java.io.File
+import java.nio.file.Files
+
+val siteDir = "www"
+
+@main def main = Pulumi.run {
+  // Create a bucket and expose a website index document
+  val siteBucket = s3.Bucket(
+    "s3-website-bucket",
+    s3.BucketArgs(
+      website = BucketWebsiteArgs(
+        indexDocument = "index.html"
+      )
+    )
+  )
+
+  val siteBucketName = siteBucket.bucket
+
+  val siteBucketPublicAccessBlock = siteBucketName.flatMap { name =>
+    s3.BucketPublicAccessBlock(
+      s"${name}-publicaccessblock",
+      s3.BucketPublicAccessBlockArgs(
+        bucket = siteBucket.id,
+        blockPublicPolicy = false // Do not block public bucket policies for this bucket
+      )
+    )
+  }
+
+  // Set the access policy for the bucket so all objects are readable
+  val siteBucketPolicy = siteBucketName.flatMap(name =>
+    s3.BucketPolicy(
+      s"${name}-access-policy",
+      s3.BucketPolicyArgs(
+        bucket = siteBucket.id,
+        policy = JsObject(
+          "Version" -> JsString("2012-10-17"),
+          "Statement" -> JsArray(
+            JsObject(
+              "Sid" -> JsString("PublicReadGetObject"),
+              "Effect" -> JsString("Allow"),
+              "Principal" -> JsObject(
+                "AWS" -> JsString("*")
+              ),
+              "Action" -> JsArray(JsString("s3:GetObject")),
+              "Resource" -> JsArray(JsString(s"arn:aws:s3:::${name}/*"))
+            )
+          )
+        ).prettyPrint
+      ),
+      CustomResourceOptions(
+        dependsOn = siteBucketPublicAccessBlock.map(List(_))
+      )
+    )
+  )
+
+  // For each file in the directory, create an S3 object stored in `siteBucket`
+  val uploads: Array[Output[BucketObject]] = File(siteDir).listFiles().map { file =>
+    s3.BucketObject(
+      NonEmptyString(file.getName).get, // FIXME: https://github.com/VirtusLab/besom/issues/138
```

Comment:
That won't get fixed by #138. `file.getName` is a completely dynamic string value coming from java.io APIs. Functional approach would be to patmat that and handle error case by failing the Output (or skipping the file, it's up to the user). I'm actually not sure if we have an Output.fail. The discussion on error handling APIs didn't occur yet. Throwing works because we catch but it's not pure.